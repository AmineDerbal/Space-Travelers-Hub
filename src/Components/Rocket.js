import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'react-bootstrap';
import Badge from 'react-bootstrap/Badge';
import { useDispatch } from 'react-redux';
import { bookRocket, cancelBooking } from '../redux/rockets/rocketsSlice';
import './styles/Rocket.css';

const Rocket = ({ rocket }) => {
  const dispatch = useDispatch();
  return (
    <div className="rocket-card">
      <img className="rocket-img" src={rocket.image} alt="rocket" />
      <div className="rocket-informations">
        <h3 className="rocket-name">{rocket.name}</h3>
        <p className="rocket-description">
          {rocket.reserved && (
          <Badge bg="success" className="reserve-badge">
            Reserved
          </Badge>
          )}
          {rocket.description}
        </p>
        {rocket.reserved && (
        <Button
          onClick={() => {
            dispatch(cancelBooking(rocket.id));
          }}
          className="reserve-rocket"
          variant="outline-secondary"
        >
          Cancel Reservation
        </Button>
        )}
        {!rocket.reserved && (
        <Button
          onClick={() => {
            dispatch(bookRocket(rocket.id));
          }}
          className="reserve-rocket"
          variant="primary"
        >
          Reserve Rocket
        </Button>
        )}

      </div>
    </div>
  );
};

Rocket.propTypes = {
  rocket: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    description: PropTypes.string,
    image: PropTypes.string,
    reserved: PropTypes.bool,
  }).isRequired,
};

export default Rocket;
