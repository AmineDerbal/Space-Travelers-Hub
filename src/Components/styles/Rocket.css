.rocket-card {
  display: flex;
  width: 90%;
  margin: 2rem auto;
  gap: 1rem;
}

.rocket-img {
  width: 250px;
  height: 200px;
}

.rocket-name {
  margin: 0;
  font-size: 1.5rem;
}

.rocket-description {
  margin: 0;
  padding-top: 0.5rem;
  font-weight: 400;
  font-size: 1.2rem;
}

.reserve-rocket {
  margin-top: 1.5rem;
}

.reserve-badge {
  margin-right: 1rem;
}
