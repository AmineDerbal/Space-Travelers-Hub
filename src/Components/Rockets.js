import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getRocketsData } from '../redux/rockets/rocketsSlice';
import Loader from './Loader';
import Rocket from './Rocket';

const Rockets = () => {
  const {
    rockets, isLoading, hasError, isFetched,
  } = useSelector((store) => store.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    if (rockets.length === 0) {
      dispatch(getRocketsData());
    }
  }, [dispatch, isFetched, rockets.length]);
  if (isLoading) {
    return <Loader />;
  }

  if (hasError) return <h2>An error has occured</h2>;
  if (rockets.length === 0) return <h2>No Rockets exist</h2>;

  return rockets.map((rocket) => <Rocket key={rocket.id} rocket={rocket} />);
};

export default Rockets;
