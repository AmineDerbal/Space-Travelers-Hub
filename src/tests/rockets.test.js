import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../redux/store';
import Rockets from '../Components/Rockets';
import Rocket from '../Components/Rocket';
import RocketTest from './rockets/rocketsReducer';
import '@testing-library/jest-dom';

const rocket = {
  rocketName: 'Heloo',
  description: 'Heloof',
  flickrImages: 'Helookn',
  reserved: false,
  id: '2',
};

describe('Rockets must pass the test', () => {
  test('added Rockets must return data', () => {
    expect(RocketTest.addedRockets()).toBeDefined();
  });
  test('added Rockets return value length must be 4', () => {
    expect(RocketTest.addedRockets()).toHaveLength(4);
  });
  test('added Rockets must return name t', () => {
    expect(RocketTest.addedRockets()[3].description).toBe('easy taske done today');
  });
});

describe('Rockets component', () => {
  it('renders Rockets component', () => {
    const root = renderer
      .create(
        <Provider store={store}>
          <Rockets />
        </Provider>,
      )
      .toJSON();
    expect(root).toMatchSnapshot();
  });
  it('renders Rocket component', () => {
    const root = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Rocket rocket={rocket} />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(root).toMatchSnapshot();
  });
});
