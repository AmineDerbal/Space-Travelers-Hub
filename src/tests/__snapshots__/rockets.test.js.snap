// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component renders Rocket component 1`] = `
<div
  className="rocket-card"
>
  <img
    alt="rocket"
    className="rocket-img"
  />
  <div
    className="rocket-informations"
  >
    <h3
      className="rocket-name"
    />
    <p
      className="rocket-description"
    >
      Heloof
    </p>
    <button
      className="reserve-rocket btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;

exports[`Rockets component renders Rockets component 1`] = `
<h2>
  No Rockets exist
</h2>
`;
